Add measurements to blocks

automata opened the pull request:
With this PR, I think we are covering the measurements we have now. It adds some new blocks that were originally undocumented, with descriptions and valid/invalid examples.

Some considerations and questions that I'd like to discuss:
- `image`: is `cover` required? I considered `cover` as part of measurement, so it's not required for now. The same is valid for the following cases:
  - `item`: are `id` and `type` required?
  - `list`: is `items` required?
  - `text`: are `text` and `length` required?
  - `video`: are `cover`, `src`, `video` and `html` required?
- `video` has `src` or just `video.src`? UPDATE: Videos do not have to have `src`, just `html`.
- `code`, `audio`, `quote` and `location` are not supported/tested on `caliper`, are we supporting them right now? If yes I need examples to add them to apidocs
- `collection` and `media` just have definitions, how can we refer to that on examples tests?
- `metadata` and `notification` don't have measurements, so this PR doesn't supports them yet

I'm aware that `saliency` and `negative_space` will change soon and then I'll need to change apidocs as well.

@bergie, @jonnor, can you guys please review? Thank you!

automata committed:
Initial drafts for contentblocks with measurements from Caliper.

automata committed:
Add media type

Media is the parent of types like image and video.

automata committed:
Add negative_space

automata committed:
Add descriptions and examples for media

automata committed:
Add examples for image type

automata committed:
Add examples for base type

automata committed:
Add examples to cta type

automata committed:
Fix typo

automata committed:
Add examples for item type

automata committed:
Add examples and description to list type

automata committed:
Add description and change requirements to text type

automata committed:
Add examples to text type

automata committed:
Add description and examples to video type

automata committed:
Add article type

automata committed:
Cover isn't required

automata committed:
Add better descriptions to image

automata committed:
Cover is not required for image

automata committed:
Add examples to article type

automata committed:
Images without cover are valid

automata committed:
Add better description to contentblock type

automata committed:
Better description to media type

automata committed:
Measurement data are not required for headline

automata committed:
H1-6 are valid headlines

automata committed:
Add descriptions to cta type

automata committed:
Add better description to headline

jonnor commented:
code, audio, quote and location are all supported, though possibly not tested. You might find some examples searching the page fixtures in our design systems.

jonnor commented:
Item generally must have type and id. One exception, it is optional when creating a new item with the API. I think we can use `required` for now, and leave a comment about that special case.

jonnor commented:
`html` is required for all blocks

jonnor commented on examples/cta.yml:
```diff
@@ -0,0 +1,165 @@
+    type: "cta"
+    uuid: "71bfc2e0-4a96-11e4-916c-0800200c9a66"
+    verb: 42
+    price: "96"
```
probably only purchase CTA will have price? @bergie

Other CTAs exists too like Follow, play video/music etc

jonnor commented on schemata/cta.yaml:
```diff
@@ -0,0 +1,34 @@
+- object
+properties: 
+  id:
+    "$ref": "base.json#/definitions/uuid"
```
I believe there is a way in JSON schema to avoid having to copy all these "base" parameters for blocks:
http://spacetelescope.github.io/understanding-json-schema/structuring.html#extending
This cleanup can be done in a second round though. We have enough testcases to be able to do what without breaking things I think :)

jonnor commented on schemata/media.yaml:
```diff
@@ -0,0 +1,1263 @@
+        description: "Extracted colors from the image, represented as an array of 5 RGB colors"
+        type: array
+        example: [[122,174,213],[8,12,21],[255,254,255],[75,108,141],[202,207,203]]
+      saliency:
```
I think big/complex data like `saliency,` `color` and `faces` should be in its own definitions, and referred to. (the property name itself should stay in media/cover). Maybe an `imagemeasurements` file? Can be done in separate pull request, but I think that would clean things up quite a bit.

jonnor commented on schemata/media.yaml:
```diff
@@ -0,0 +1,1263 @@
+                        type: integer
+                        example: 1803
+              -
+                type: object
```
Whats with the 5 times copy of the whole x,y,width,height,text thing? That can't be right!?

jonnor commented on schemata/media.yaml:
```diff
@@ -0,0 +1,1263 @@
+                      -
+                        type: integer
+                  -
+                    type: array
```
Same duplicate/copy/paste issue here, just worse. Needs fixing.

jonnor commented:
@automata the duplication thing mentioned in last two comments must be fixed before this goes in. The other things can/should be done in a later pull request.

jonnor commented:
@automata any update to this? Bunch of comments above

automata commented:
Agree with the comments, thank you! I just have to find a time to finish this stuff.

jonnor commented:
@automata ping

jonnor commented:
@automata ping

automata committed:
Bump grunt-aglio dependency

automata committed:
Add geographic data to base type

automata committed:
Add geographic data to metadata

automata committed:
Add location schema

automata committed:
Add examples of location

automata committed:
Add audio schema

automata committed:
Add code schema

automata committed:
Add quote schema

automata committed:
Add examples of quote

automata committed:
Add examples of code

automata committed:
Add examples to audio schema

automata committed:
Add examples to location schema

automata committed:
Update audio schema

automata committed:
Add example of a Follow CtA

automata committed:
Remove price requirement on cta schema

automata committed:
Add schema for complex image measurements

automata committed:
Start using imagemeasurements schema on media

automata committed:
Validate faces array on media schema

automata committed:
Remove negative_space property from saliency

automata committed:
Add geometric elements as base definitions

automata committed:
Add saliency property

automata committed:
Reference to saliency sub-schema

automata committed:
Add histogram and exif sub-schemas

automata committed:
Update media schema with latest features

automata commented:
I think that's the most updated version of all measurements/features we're extracting. I considered all the suggestions from @jonnor, removing duplications and adding missing schemas like `quote`, `location`, `code` and `audio`.

Price property on `cta` is not required anymore and example of "follow" CtA was created.

I think there's only two points missing:
1. I guess we should have all properties/schemas we calculate on data-helper added to `media/cover` schema. Maybe we can create a `datahelper` schema and use `allOf` on `media/cover` to extend it. At the same time we will make it explicit that we get such data from data-helper
2. Review everything, specially improving descriptions and checking required properties

@bergie @cbchouinard @jonnor suggestions?

automata committed:
Add valid example of image with measurements from Caliper

automata committed:
Only id, type and html are required

automata committed:
Add definitions from data-helper measurements

All measurements calculated at a post-analysis step
(by data-helper) are optional and extends `media` schema.

They are defined in its own schema.

automata committed:
Describe EXIF definition with more details

automata committed:
Extend media schema with data-helper measurements

automata committed:
Add scene to image example

automata commented:
All measurements calculated on data-helper are also included as definitions on `helpermeasurements` schema. It's included on `media` schema.

automata committed:
Set src as a definition

automata committed:
Require src and html

automata committed:
Require src and html

automata committed:
Warn about deprecation

automata committed:
Tweak descriptions

automata committed:
Add examples

automata committed:
Require src on image schema

automata committed:
Tweak descriptions a bit

automata committed:
Require src and html on video

automata commented:
@bergie it's probably good for a review.

jonnor commented on schemata/base.yaml:
```diff
@@ -28,3 +28,114 @@ definitions:
+        type: number
+        description: Zoom level of map
+        example: 4.0
+  bbox:
```
Isn't this just `rectangle` ? `bbox` is something only image nerds understand what is.

automata replied:
`bbox` is only used on image measurements for now, for faces, salient regions and negative spaces, so IMHO it's on context.

jonnor commented on schemata/base.yaml:
```diff
@@ -28,3 +28,114 @@ definitions:
+        type: number
+        description: "Height"
+        example: 229.2087
+  tuplecolor:
```
Better to call `colorarray` or even `rgbarray` ? `description should mention this is using 8 bit integer colors`[0 255]`and not`[0.0 1.0]`.

automata replied:
Yes, much better!

jonnor commented on schemata/contentblock.yaml:
```diff
@@ -39,13 +34,29 @@ properties:
     type: [number, "null"]
-  src:
-    type: [string, "null"]
+  anyOf: [headline, text, cta, media, list]
```
Nice!

jonnor commented on schemata/imagemeasurements.yaml:
```diff
@@ -0,0 +1,216 @@
+        example: [[329,238],[539,372]]
+  histogram:
+    description: "Histogram of color levels"
+    type: object
```
Isn't `histogram` an `array` not object?

automata replied:
`histogram` is an `object` with `array` properties (`histogram.r`, `histogram.l`, ...).

jonnor commented:
Merged since this is huge and looks mostly correct. The few comments above can be addressed in another PR

automata commented:
@jonnor agreed, thanks!
